src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Controller;
+
+use Psr\Container\ContainerInterface;
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Symfony\Component\DependencyInjection\ServiceSubscriberInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Templating\EngineInterface;
+
+/**
+ * Provides common features needed in controllers.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+abstract class AbstractController implements ServiceSubscriberInterface
+{
+    use ControllerTrait;
+
+    private $container;
+
+    public function setContainer(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    public static function getSubscribedServices()
+    {
+        return array(
+            'router' => '?'.RouterInterface::class,
+            'request_stack' => '?'.RequestStack::class,
+            'http_kernel' => '?'.HttpKernelInterface::class,
+            'serializer' => '?'.SerializerInterface::class,
+            'session' => '?'.SessionInterface::class,
+            'security.authorization_checker' => '?'.AuthorizationCheckerInterface::class,
+            'templating' => '?'.EngineInterface::class,
+            'twig' => '?'.\Twig_Environment::class,
+            'doctrine' => '?'.ManagerRegistry::class,
+            'form.factory' => '?'.FormFactoryInterface::class,
+            'security.token_storage' => '?'.TokenStorageInterface::class,
+            'security.csrf.token_manager' => '?'.CsrfTokenManagerInterface::class,
+        );
+    }
+}
```

Comment:
It's nice that `Controller` and `AbstractController` are sharing code (`ControllerTrait`), but there may be some advantages to duplicating the functions in each. Specifically, error messages. Here are some situations with `AbstractController` where we should give a really clear error:

1) I don't understand things, and extend `AbstractController` on a non-service controller -> `setContainer()` won't be called (and currently, we'll get a "method called on a non-object" error)
2) I don't have the `_instanceof` for `AbstractController` or I'm missing the calls part. Same error as above.
3) All of the dependencies in `AbstractController` are marked as optional. If any of these are truly optional, then the methods that use these services should check if they exist, and throw a clear exception if those methods are called, but the service is missing.